src/libAtomVM/globalcontext.c

```diff
@@ -135,6 +135,26 @@ void globalcontext_register_process(GlobalContext *glb, int atom_index, int loca
     linkedlist_append(&glb->registered_processes, &registered_process->registered_processes_list_head);
 }
 
+void globalcontext_unregister_process(GlobalContext *glb, int atom_index)
+{
+    if (!glb->registered_processes) {
+        return;
+    }
+
+    struct RegisteredProcess *registered_processes = GET_LIST_ENTRY(glb->registered_processes, struct RegisteredProcess, registered_processes_list_head);
+
+    struct RegisteredProcess *p = registered_processes;
+
+    do {
+        if (p->atom_index == atom_index) {
+            linkedlist_remove(&glb->registered_processes, &p->registered_processes_list_head);
+            free(p);
```

Comment:
I think this would lead to a crash, since you are going to do `free` to an address that is inside a block of allocated memory, not the block itself that you want to free.

Reply:
I will go back and look at that... but strangely it does not crash, and functions as expected. several hundred starts and stops does not seem to cause any memory leak when comparing esp32 available memory with `erlang:system_info(esp32_free_heap_size)` after each `unregister`.

Reply:
I wanted to test this so I added:
`fprintf(stderr, "Allocated memory at: %p.\n", (void *)registered_process);`
after the malloc in `globalcontext_register_process` and:
`fprintf(stderr, "Free memory at: %p.\n", (void *)p);`
before `free(p);` in `globalcontext_unregister_process`

Running a test app to register and then unregister a process logs the following:

About to register(test, <0.2.0>)...
Allocated memory at: 0x5555555f0b80.
About to unregister(test)...
Free memory at: 0x5555555f0b80.
Return value: ok

So I really do believe this is correct. I think my choice to use the same variable name `registered_processes`, as in `globalcontext_register_process` makes it a little misleading. Perhaps I should change the name to `registered_process` to make it more clear?

Reply:
At the moment you are right, but it is not true in general, I will explain here why:
`struct RegisteredProcess` is defined as:
```c
struct RegisteredProcess
{
    struct ListHead registered_processes_list_head;

    int atom_index;
    int local_process_id;
};
```

Therefore usually `registered_processes_list_head` and and the `struct RegisteredProcess` end having the same address in memory, since `registered_processes_list_head` has offset 0.
However:
1. the code is quite fragile, if anyone at line 34 adds something before `registered_processes_list_head` or reorders the struct, the function at line 160 will stop working
2. I'm not even sure if standard C requires the first member to be at offset 0, or if any kind of arbitrary padding is allowed, hence it might work differently with different compilers or on different architectures (or even with different compiler flags)

Reply:
@bettio Can you suggest an alternative?  I can't really think of anything except something awkward like
```
struct RegisteredProcess
{
    struct ListHead registered_processes_list_head;

    int atom_index;
    int local_process_id;
    void *self;
};
```
where `self` points to the struct, itself.  So you could do something like
```
free(p->self);
```
But that is wasting a whole word of memory.

The spec is probably (deliberately) ambiguous:

> _A structure type describes a sequentially allocated nonempty set of member objects (and, in certain circumstances, an incomplete array), each of which has an optionally specified name and possibly distinct type._

Contrast that with an array type, which is "contiguously allocated"...

Reply:
sorry, thank to all of you about getting back to this, I must have been tired and I made a huge mistake here: my first comment did not make sense, and I also explained the reason why it doesn't make sense just on my next comment.
I really apologize.